### PR TITLE
Grouping Mesh methods in Doxygen [mesh-group-doc-dev]

### PR DESCRIPTION
**Pull request by mlstowell**
This PR builds upon PR #3131.

This PR should not change any functionality of the `Mesh` class. The intention was only to rearrange method declarations so that Doxygen could organize the numerous class methods into more manageable subsets. I tried to leave the method documentation unchanged and only rearrange the existing descriptions.

All public methods in the `Mesh` class should now be in a Doxygen "member group". Clearly there could be different choices for defining these groups and I am happy to hear other people's thoughts and suggestions. My hope would be to cross reference the groups so that users can more easily find the methods they are looking for even if they don't appear in the group that they might have expected.  

I added `@todo` declarations in a handful of new Doxygen comments. Doxygen uses these to generate a separate page with a todo list. I did this for two reasons; firstly I was unsure how to describe these methods and how to categorize them, secondly I was hoping other team members might offer to document these methods and such a list seemed a very convenient tool.

 | PR | Author | Editor | Reviewers | Assignment | Approval | Merge|
 | --- | --- | --- | --- | --- | --- | --- |
| [#3217](https://github.com/mfem/mfem/pull/3217) | @mlstowell | @tzanio | @atallah727 + @acfisher | 6/6/23 | 7/14/23 | 7/18/23 | |

<details>
<summary>PR Checklist</summary>
    
- [ ] Code builds.
- [ ] Code passes `make style`.
- [ ] Update `CHANGELOG`:
    - [ ] Is this a new feature users need to be aware of? New or updated example or miniapp?
    - [ ] Does it make sense to create a new section in the `CHANGELOG` to group with other related features?
- [ ] Update `INSTALL`:
    - [ ] Had a new optional library been added? If so, what range of versions of this library are required? (*Make sure the external library is compatible with our BSD license, e.g. it is not licensed under GPL!*)
    - [ ] Have the version ranges for any required or optional libraries changed?
    - [ ] Does `make` or `cmake` have a new target?
    - [ ] Did the requirements or the installation process change? *(rare)*
- [ ] Update continuous integration server configurations if necessary (e.g. with new version requirements for each of MFEM's dependencies)
    - [ ] `.github`
    - [ ] `.appveyor.yml`
- [ ] Update `.gitignore`:
    - [ ] Check if `make distclean; git status` shows any files that were generated from the source by the project (not an IDE) but we don't want to track in the repository.
    - [ ] Add new patterns (just for the new files above) and re-run the above test.
- [ ] New examples:
    - [ ] All sample runs at the top of the example source file work.
    - [ ] Update `examples/makefile`:
      - [ ] Add the example code to the appropriate `SEQ_EXAMPLES` and `PAR_EXAMPLES` variables.
      - [ ] Add any files generated by it to the `clean` target.
      - [ ] Add the example binary and any files generated by it to the top-level `.gitignore` file.
    - [ ] Update `examples/CMakeLists.txt`:
      - [ ] Add the example code to the `ALL_EXE_SRCS` variable.
      - [ ] Make sure `THIS_TEST_OPTIONS` is set correctly for the new example.
   - [ ] List the new example in `doc/CodeDocumentation.dox`.
   - [ ] If new examples directory (e.g.`examples/pumi`), list it in `doc/CodeDocumentation.conf.in`
   - [ ] Companion pull request for documentation in [mfem/web](https://github.com/mfem/web) repo:
      - [ ] Update or add example-specific documentation, see e.g. the `src/examples.md`.
      - [ ] Add the description, labels and screenshots in `src/examples.md` and `src/img`.
      - [ ] In `examples.md`, list the example under the appropriate categories, add new categories if necessary.
      - [ ] Add a short description of the example in the "Extensive Examples" section of `features.md`.
- [ ] New miniapps:
   - [ ] All sample runs at the top of the miniapp source file work.
   - [ ] Update top-level `makefile` and `makefile` in corresponding miniapp directory.
   - [ ] Add the miniapp binary and any files generated by it to the top-level `.gitignore` file.
   - [ ] Update CMake build system:
      - [ ] Update the `CMakeLists.txt` file in the `miniapps` directory, if the new miniapp is in a new directory.
      - [ ] Add/update the `CMakeLists.txt` file in the new miniapp directory.
      - [ ] Consider adding a new test for the new miniapp.
   - [ ] List the new miniapp in `doc/CodeDocumentation.dox`
   - [ ] If new miniapps directory (e.g.`miniapps/nurbs`), add it to `MINIAPP_SUBDIRS` in the `makefile`.
   - [ ] If new miniapps directory (e.g.`miniapps/nurbs`), list it in `doc/CodeDocumentation.conf.in`
   - [ ] Companion pull request for documentation in [mfem/web](https://github.com/mfem/web) repo:
     - [ ] Update or add miniapp-specific documentation, see e.g. the `src/meshing.md` and `src/electromagnetics.md` files.
     - [ ] Add the description, labels and screenshots in `src/examples.md` and `src/img`.
     - [ ] The miniapps go at the end of the page, and are usually listed only under a specific "Application (PDE)" category.
     - [ ] Add a short description of the miniapp in the "Extensive Examples" section of `features.md`.
- [ ] New capability:
   - [ ] All new public, protected, and private classes, methods, data members, and functions have full Doxygen-style documentation in source comments. Documentation should include descriptions of member data, function arguments and return values, template parameters, and prerequisites for calling new functions.
   - [ ] Pointer arguments and return values must specify whether ownership is being transferred or lent with the call.
   - [ ] Any new functions should include descriptions of their intended use e.g. for internal use only, user-facing, etc., along with references to example code whenever possible/appropriate.
   - [ ] Consider adding new sample runs in existing examples to highlight the new capability.
   - [ ] Consider saving cool simulation pictures with the new capability in the Confluence gallery (LLNL only) or submitting them, via pull request, to the gallery section of the `mfem/web` repo.
   - [ ] If this is a major new feature, consider mentioning it in the short summary inside `README` *(rare)*.
   - [ ] List major new classes in `doc/CodeDocumentation.dox` *(rare)*.
- [ ] Update this checklist, if the new pull request affects it.
- [ ] Run `make unittest` to make sure all unit tests pass.
- [ ] Run the tests in `tests/scripts`.
- [ ] (LLNL only) After merging:
   - [ ] Update internal tests to include the new features.
</details>

### CODE REVIEW

**Comment by acfisher** [comment]
Is this getting close to ready for review?

**Comment by mlstowell** [comment]
I believe this was waiting for PR #3131 which has been merged for a while now. I'll revisit this one and get it ready asap.

**Comment by tzanio** [comment]
This PR is now [under review](https://github.com/mfem/mfem/blob/master/CONTRIBUTING.md#pull-requests) (see the table in the PR description). To help with the review process, please **do not force push** to the branch.

**Comment by tzanio** [comment]
@mlstowell, can you please resolve the conflicts?

File: mesh/mesh.hpp
Comment (acfisher) [inline mesh/mesh.hpp:597]: Does the user need to call anything after they change the vertices?

File: mesh/mesh.hpp
Comment (acfisher) [inline mesh/mesh.hpp:638]: Maybe put a link here to the mentioned mesh constructor.

File: mesh/mesh.hpp
Comment (acfisher) [inline mesh/mesh.hpp:1000]: May want to mention that this can be bigger than the dimension for eg. surface meshes in 3D.

File: mesh/mesh.hpp
Comment (acfisher) [inline mesh/mesh.hpp:1004]: Should we call this EulerNumber3D?  Should we error if dimension is wrong?

File: mesh/mesh.hpp
Comment (acfisher) [inline mesh/mesh.hpp:1035]: Does this count boundary elements?

File: mesh/mesh.hpp
Comment (acfisher) [inline mesh/mesh.hpp:1046]: Might want to mention that with high order meshes it will be important to have enough refinement otherwise the bounding box may clip a bump.

File: mesh/mesh.hpp
Comment (acfisher) [inline mesh/mesh.hpp:1050]: It may be worth a comment here on what each of these characteristics are.

File: mesh/mesh.hpp
Comment (acfisher) [inline mesh/mesh.hpp:1232]: What does type do here?  Do we have the specify the element type in this method?

File: mesh/mesh.hpp
Comment (acfisher) [inline mesh/mesh.hpp:1236]: What is this dir vector about here?

File: mesh/mesh.hpp
Comment (acfisher) [inline mesh/mesh.hpp:1734]: Could probably use an explanation of what the point matrix is here.  Which points.  Are they in the columns or rows of the matrix?

File: mesh/mesh.hpp
Comment (acfisher) [inline mesh/mesh.hpp:1967]: Doxygenify

**Comment by acfisher** [comment]
This is a huge improvement to our Mesh documentation.  I put in a few minor to moderate comments to help clarify a few things.  If any of that looks like it will be too much work, perhaps todo it and we can pull this good work in.

**Comment by mlstowell** [comment]
@acfisher, thanks for the thorough review and the excellent comments. I did want to keep this PR focused on re-grouping the methods to make related functionality easier to find. I did not want to change any of the existing comments in this PR.

That being said, you've made very good suggestions and I don't want your review to go unaddressed. If it's okay with you and @tzanio I'll write a quick follow-up PR to make your suggested changes after this PR is merged.

File: mesh/mesh.hpp
Comment (mlstowell) [reply mesh/mesh.hpp:597]: No, but I can expand on this documentation to make this more clear.

**Comment by acfisher** [comment]
> @acfisher, thanks for the thorough review and the excellent comments. I did want to keep this PR focused on re-grouping the methods to make related functionality easier to find. I did not want to change any of the existing comments in this PR.
> 
> That being said, you've made very good suggestions and I don't want your review to go unaddressed. If it's okay with you and @tzanio I'll write a quick follow-up PR to make your suggested changes after this PR is merged.

I thin kthat's a great idea.  Let's push this one forward.

**Review by acfisher — APPROVED** [review]
Thanks for this excellent doxygen work.

**Review by tzanio — APPROVED** [review]

**Comment by tzanio** [comment]
Merged in `next` for testing...

**Pull request merged**
